bindings/cs/Protean-cs/Protean.Net/BinaryWriter.cs

```diff
@@ -65,357 +65,561 @@ public static byte[] ToBytes(Variant value, BinaryMode mode)
             }
-		}
-
-		public static byte[] ToBytes(Variant value)
-		{
-			return ToBytes(value, BinaryMode.Default);
-		}
-
-		public void Write(Variant v)
-		{
-			WriteHeader();
-			WriteVariant(v);
-		}
-
-		protected void WriteHeader()
-		{
-			// Create and write the Protean header
-			m_stream.Write(BitConverter.GetBytes(BinaryConstants.PROTEAN_MAGIC), 0, 4);
-			m_stream.Write(BitConverter.GetBytes(BinaryConstants.VERSION_MAJOR << 16 | BinaryConstants.VERSION_MINOR), 0, 4);
-			m_stream.Write(BitConverter.GetBytes((int)m_mode), 0, 4);
-		}
-
-		protected void Write(string arg)
-		{
-			int length = arg.Length;
-			byte[] bytes = new byte[length];
-
-			for (int i = 0; i < length; ++i)
-			{
-				bytes[i] = (byte)arg[i];
-			}
-
-			Write(length);
-			Write(bytes, true);
-		}
-
-		protected void Write(TimeSpan arg)
-		{
-			Write((Int64)arg.TotalMilliseconds);
-		}
-
-		private static readonly long MaxDateTimeMillis = (Variant.MaxDateTime.Ticks - Variant.MinDateTime.Ticks) / 10000;
-
-		protected void Write(DateTime arg)
-		{
-			if (arg == DateTime.MaxValue)
-			{
-				Write(MaxDateTimeMillis);
-			}
-			else
-			{
-				Write((arg.Ticks - Variant.MinDateTime.Ticks) / 10000);
-			}
-		}
-
-		private void Write(byte arg)
-		{
-			m_filter.WriteByte(arg);
-		}
-
-		protected void Write(Int32 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(UInt32 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(Int64 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(UInt64 arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(float arg)
-		{
-			Write(BitConverter.GetBytes(arg));
-		}
-
-		protected void Write(double arg)
-		{
-			Write(BitConverter.DoubleToInt64Bits(arg));
-		}
-
-		private void Write(byte[] arg)
-		{
-			m_filter.Write(arg, 0, arg.Length);
-		}
-
-		protected void Write(bool arg)
-		{
-			Write((arg ? 1 : 0));
-		}
-
-		protected void Write(byte[] bytes, bool writePadding)
-		{
-			Write(bytes);
-
-			if (writePadding)
-			{
-				int residual = (4 - (bytes.Length % 4)) % 4;
-				for (int i = 0; i < residual; ++i)
-				{
-					Write((byte)0);
-				}
-			}
-		}
-
-	    protected void Write(DataTable arg)
-	    {
+        }
+
+        public static byte[] ToBytes(Variant value)
+        {
+            return ToBytes(value, BinaryMode.Default);
+        }
+
+        public void Write(Variant v)
+        {
+            WriteHeader();
+            WriteVariant(v);
+        }
+
+        public static int GetBytes(Variant value, BinaryMode mode, byte[] buffer, int bufferIndex)
+        {
+            using (MemoryStream ms = new MemoryStream(buffer, bufferIndex, buffer.Length - bufferIndex))
+            {
+                using (BinaryWriter writer = new BinaryWriter(ms, mode))
+                {
+                    writer.Write(value);
+                }
+
+                return (int)ms.Position;
```

Comment:
Is there any reason to manipulate the type here, rather than using the more natural `long` throughout these methods?

(I realise we are unlikely to be writing very many 2GB variants...)

Reply:
The MemoryBuffer class is inconsistent. The offset/index you can pass in is an int, not sure why Position would be a long then. I mainly cast it here so the client doesn't have to do it. Also arrays with indexes > max int are quite hard to deal with. You need to use Array.GetItem() instead of the bracket indexers.